commit/chainfee/outcome.go

```diff
@@ -0,0 +1,124 @@
+package chainfee
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugincommon"
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugincommon/consensus"
+)
+
+func (p *processor) getConsensusObservation(
+	aos []plugincommon.AttributedObservation[Observation],
+) (Observation, error) {
+	aggObs := aggregateObservations(aos)
+
+	// consensus on the fChain map uses the role DON F value
+	// because all nodes can observe the home chain.
+	donThresh := consensus.MakeConstantThreshold[cciptypes.ChainSelector](consensus.TwoFPlus1(p.fRoleDON))
+	fChains := consensus.GetConsensusMap(p.lggr, "fChain", aggObs.FChain, donThresh)
+
+	fDestChain, exists := fChains[p.destChain]
+	if !exists {
+		return Observation{},
+			fmt.Errorf("no consensus value for fDestChain, destChain: %d", p.destChain)
+	}
+	if len(aggObs.Timestamps) < fDestChain {
+		return Observation{},
+			fmt.Errorf("not enough observations for timestamps to reach consensus, have %d, need %d",
+				len(aggObs.Timestamps), fDestChain)
+	}
+	timestamp := consensus.Median(aggObs.Timestamps, consensus.TimestampComparator)
+
+	chainFeeUpdatesConsensus := consensus.GetConsensusMapAggregator(
+		p.lggr,
+		"ChainFeeLatestUpdates",
+		aggObs.ChainFeeLatestUpdates,
+		consensus.MakeConstantThreshold[cciptypes.ChainSelector](consensus.TwoFPlus1(fDestChain)),
+		consensus.TimestampMedianAggregator,
+	)
+
+	twoFPlus1 := consensus.MakeMultiThreshold(fChains, consensus.TwoFPlus1)
+
+	feeComponents := consensus.GetConsensusMapAggregator(
+		p.lggr,
+		"FeeComponents",
+		aggObs.FeeComponents,
+		twoFPlus1,
+		// Aggregator function
+		func(vals []types.ChainFeeComponents) types.ChainFeeComponents {
+			executionFees := make([]cciptypes.BigInt, len(vals))
+			dataAvailabilityFees := make([]cciptypes.BigInt, len(vals))
+			for i, feeComp := range vals {
+				executionFees[i] = cciptypes.NewBigInt(feeComp.ExecutionFee)
+				dataAvailabilityFees[i] = cciptypes.NewBigInt(feeComp.DataAvailabilityFee)
+			}
+			return types.ChainFeeComponents{
+				ExecutionFee:        consensus.Median(executionFees, consensus.BigIntComparator).Int,
+				DataAvailabilityFee: consensus.Median(dataAvailabilityFees, consensus.BigIntComparator).Int,
```

Comment:
This doesn't handle overflows

Reply:
I believe `Median` does a `sort` then picks out the middle element, there's no addition going on.

Reply:
It's because I'm using `cciptypes.BigInt`, to get the `*big.Int` I call `.Int`. TBH this cciptypes.BigInt is annoying and I'd rather use only `*big.Int` directly where possible.